In [2]:
text = '''> **Physician:** *Good morning, Ms. Jones. How are you feeling today?*
**Patient:** *Good morning, doctor. I’m doing better, but I still have some discomfort now and then.*

 **Physician:** *I understand you were in a car accident last September. Can you walk me through what happened?*

**Patient:** *Yes, it was on September 1st, around 12:30 in the afternoon. I was driving from Cheadle Hulme to Manchester when I had to stop in traffic. Out of nowhere, another car hit me from behind, which pushed my car into the one in front.*
>
> **Physician:** *That sounds like a strong impact. Were you wearing your seatbelt?*
>
> **Patient:** *Yes, I always do.*
>
> **Physician:** *What did you feel immediately after the accident?*
>
> **Patient:** *At first, I was just shocked. But then I realized I had hit my head on the steering wheel, and I could feel pain in my neck and back almost right away.*
>
> **Physician:** *Did you seek medical attention at that time?*
>
> **Patient:** *Yes, I went to Moss Bank Accident and Emergency. They checked me over and said it was a whiplash injury, but they didn’t do any X-rays. They just gave me some advice and sent me home.*
>
> **Physician:** *How did things progress after that?*
>
> **Patient:** *The first four weeks were rough. My neck and back pain were really bad—I had trouble sleeping and had to take painkillers regularly. It started improving after that, but I had to go through ten sessions of physiotherapy to help with the stiffness and discomfort.*
>
> **Physician:** *That makes sense. Are you still experiencing pain now?*
>
> **Patient:** *It’s not constant, but I do get occasional backaches. It’s nothing like before, though.*
>
> **Physician:** *That’s good to hear. Have you noticed any other effects, like anxiety while driving or difficulty concentrating?*
>
> **Patient:** *No, nothing like that. I don’t feel nervous driving, and I haven’t had any emotional issues from the accident.*
>
> **Physician:** *And how has this impacted your daily life? Work, hobbies, anything like that?*
>
> **Patient:** *I had to take a week off work, but after that, I was back to my usual routine. It hasn’t really stopped me from doing anything.*
>
> **Physician:** *That’s encouraging. Let’s go ahead and do a physical examination to check your mobility and any lingering pain.*
>
> [**Physical Examination Conducted**]
>
> **Physician:** *Everything looks good. Your neck and back have a full range of movement, and there’s no tenderness or signs of lasting damage. Your muscles and spine seem to be in good condition.*
>
> **Patient:** *That’s a relief!*
>
**Physician:** *Yes, your recovery so far has been quite positive. Given your progress, I’d expect you to make a full recovery within six months of the accident. There are no signs of long-term damage or degeneration.*

**Patient:** *That’s great to hear. So, I don’t need to worry about this affecting me in the future?*

**Physician:** *That’s right. I don’t foresee any long-term impact on your work or daily life. If anything changes or you experience worsening symptoms, you can always come back for a follow-up. But at this point, you’re on track for a full recovery.*

**Patient:** *Thank you, doctor. I appreciate it.*

**Physician:** *You’re very welcome, Ms. Jones. Take care, and don’t hesitate to reach out if you need anything.*'''

In [3]:
import re
def clean_text(text_lines):
  cleaned_lines = []
  for line in text_lines:
    cleaned_line = re.sub(r'[>*]', '', line)
    cleaned_line = re.sub(r'\s+', ' ', cleaned_line).strip()
    cleaned_lines.append(cleaned_line)
  return cleaned_lines


In [4]:
clean_text_lines = clean_text(text.split())
print(clean_text_lines)

['', 'Physician:', 'Good', 'morning,', 'Ms.', 'Jones.', 'How', 'are', 'you', 'feeling', 'today?', 'Patient:', 'Good', 'morning,', 'doctor.', 'I’m', 'doing', 'better,', 'but', 'I', 'still', 'have', 'some', 'discomfort', 'now', 'and', 'then.', 'Physician:', 'I', 'understand', 'you', 'were', 'in', 'a', 'car', 'accident', 'last', 'September.', 'Can', 'you', 'walk', 'me', 'through', 'what', 'happened?', 'Patient:', 'Yes,', 'it', 'was', 'on', 'September', '1st,', 'around', '12:30', 'in', 'the', 'afternoon.', 'I', 'was', 'driving', 'from', 'Cheadle', 'Hulme', 'to', 'Manchester', 'when', 'I', 'had', 'to', 'stop', 'in', 'traffic.', 'Out', 'of', 'nowhere,', 'another', 'car', 'hit', 'me', 'from', 'behind,', 'which', 'pushed', 'my', 'car', 'into', 'the', 'one', 'in', 'front.', '', '', 'Physician:', 'That', 'sounds', 'like', 'a', 'strong', 'impact.', 'Were', 'you', 'wearing', 'your', 'seatbelt?', '', '', 'Patient:', 'Yes,', 'I', 'always', 'do.', '', '', 'Physician:', 'What', 'did', 'you', 'feel', '

In [5]:
a = "That’s good to hear. Have you noticed any other effects, like"
for i in a.split():
  print(i)

That’s
good
to
hear.
Have
you
noticed
any
other
effects,
like


In [6]:
new_text = " ".join(clean_text_lines)

In [7]:
!pip install https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.5.1/en_core_sci_md-0.5.1.tar.gz

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 120.2/120.2 MB 7.9 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 153.9/153.9 kB 7.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.4/6.4 MB 41.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 47.3/47.3 kB 750.3 kB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 15.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 57.0/57.0 kB 3.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 917.4/917.4 kB 38.8 MB/s eta 0:00:00
  Created wheel for en_core_sci_md: filename=en_core_sci_md-0.5.1-py3-none-any.whl size=120253138 sha256=218643d5a833c291dfe51d196d03ed61a8bccd2a0851e5d5922053332d339fe1
  Stored in directory: /root/.cache/pip/wheels/0a/50/82/7547d452aa8d5a653fb1271c38113de20f7842effc4b7313d0
Successfully built en_core_sci_md
  Attempting uninstall: wasabi
    Found existing installation: wasabi 1.1.3
    Uninstalli

In [8]:
import spacy
nlp = spacy.load("en_core_sci_md")
text = new_text
doc = nlp(text)
for ent in doc.ents:
  print(ent.text,ent.label_)

Good morning ENTITY
Ms. Jones ENTITY
Patient ENTITY
Good morning ENTITY
doctor ENTITY
discomfort ENTITY
Physician ENTITY
understand ENTITY
car ENTITY
accident ENTITY
September ENTITY
walk ENTITY
Patient ENTITY
afternoon ENTITY
Cheadle Hulme to Manchester ENTITY
stop ENTITY
traffic ENTITY
car ENTITY
car ENTITY
Physician ENTITY
sounds ENTITY
impact ENTITY
wearing ENTITY
seatbelt ENTITY
Patient ENTITY
Physician ENTITY
accident ENTITY
Patient ENTITY
I ENTITY
shocked ENTITY
realized I ENTITY
head ENTITY
steering wheel ENTITY
I ENTITY
pain ENTITY
neck ENTITY
Physician ENTITY
medical attention ENTITY
time ENTITY
Patient ENTITY
Emergency ENTITY
whiplash injury ENTITY
X-rays ENTITY
advice ENTITY
sent ENTITY
Physician ENTITY
progress ENTITY
Patient ENTITY
weeks ENTITY
rough ENTITY
neck ENTITY
back pain ENTITY
trouble sleeping ENTITY
painkillers ENTITY
regularly ENTITY
improving ENTITY
sessions ENTITY
physiotherapy ENTITY
stiffness ENTITY
discomfort ENTITY
Physician ENTITY
pain ENTITY
Patient ENT

In [9]:
from transformers import pipeline
summarizer = pipeline("summarization", model="facebook/bart-large-cnn")
summary = summarizer(new_text, max_length=100, min_length=30, do_sample=False)
print(summary[0]['summary_text'])

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/1.58k [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.63G [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/363 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/899k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

Device set to use cuda:0


Ms. Jones suffered a whiplash injury in a car accident last September. She had to go through ten sessions of physiotherapy to help with the stiffness and discomfort. There are no signs of long-term damage or degeneration.


**Summary Creation**


SUMMARY = 'Ms. Jones suffered a whiplash injury in a car accident last September. She had to go through ten sessions of physiotherapy to help with the stiffness and discomfort. There are no signs of long-term damage or degeneration.'

In [10]:
!pip install keybert

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 41.4/41.4 kB 3.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 4.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 41.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 16.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 30.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 1.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 7.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 14.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 8.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.5/207.5 MB 1.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.1/21.1 MB 68.9 MB/s eta 0:00:00
  Attempting uninstall: nvidia-nvjitlink-cu12
    Found existing installation: nvidia-nvjitlink-cu12 12.5.82
    Uninstalling

In [11]:
!pip install -U pydantic

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 431.7/431.7 kB 15.0 MB/s eta 0:00:00
  Attempting uninstall: pydantic
    Found existing installation: pydantic 1.10.21
    Uninstalling pydantic-1.10.21:
      Successfully uninstalled pydantic-1.10.21
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
spacy 3.4.4 requires pydantic!=1.8,!=1.8.1,<1.11.0,>=1.7.4, but you have pydantic 2.10.6 which is incompatible.
en-core-web-sm 3.7.1 requires spacy<3.8.0,>=3.7.2, but you have spacy 3.4.4 which is incompatible.


In [13]:
from keybert import KeyBERT
kw_model = KeyBERT()
keywords = kw_model.extract_keywords(new_text, top_n=5)
print(keywords)

modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/10.5k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

[('accident', 0.4232), ('whiplash', 0.419), ('injury', 0.3665), ('discomfort', 0.3177), ('recovery', 0.3021)]


**KEY WORD EXTRACTION**

Key_words = [('accident', 0.4232), ('whiplash', 0.419), ('injury', 0.3665), ('discomfort', 0.3177), ('recovery', 0.3021)]

In [14]:
def generate_structure_summary(text):

    subjective_part = text.split(".")[0]
    objective_part = text.split(".")[1]
    assessment_part = text.split(".")[2]
    plan_part = text.split(".")[3]
    subjective_summary = summarizer(subjective_part, max_length=50, min_length=5, do_sample=False)[0]['summary_text']
    objective_summary = summarizer(objective_part, max_length=50, min_length=2, do_sample=False)[0]['summary_text']
    assessment_summary = summarizer(assessment_part, max_length=50, min_length=5, do_sample=False)[0]['summary_text']
    plan_summary = summarizer(plan_part, max_length=50, min_length=5, do_sample=False)[0]['summary_text']
    soap_summary = f"""
    SOAP Notes:
    Subjective: {subjective_summary}
    Objective: {objective_summary}
    Assessment: {assessment_summary}
    Plan: {plan_summary}
    """

    return soap_summary


s_summary = generate_structure_summary(new_text)
print(s_summary)

Your max_length is set to 50, but your input_length is only 10. Since this is a summarization task, where outputs shorter than the input are typically wanted, you might consider decreasing max_length manually, e.g. summarizer('...', max_length=5)
Your max_length is set to 50, but your input_length is only 4. Since this is a summarization task, where outputs shorter than the input are typically wanted, you might consider decreasing max_length manually, e.g. summarizer('...', max_length=2)
Your max_length is set to 50, but your input_length is only 15. Since this is a summarization task, where outputs shorter than the input are typically wanted, you might consider decreasing max_length manually, e.g. summarizer('...', max_length=7)
Your max_length is set to 50, but your input_length is only 19. Since this is a summarization task, where outputs shorter than the input are typically wanted, you might consider decreasing max_length manually, e.g. summarizer('...', max_length=9)



    SOAP Notes:
    Subjective: The patient is seen by a doctor in the back of a waiting room. The doctor says, "Good morning, Ms.
    Objective: Mr. Jones is the son of former U.S. President Harry S. Jones. Mr. Jones has been married to his wife, Barbara, for more than 40 years.
    Assessment: Patient: Good morning, doctor. Doctor: How are you feeling today? Patient: I'm fine.
    Plan: "I’m doing better, but I still have some discomfort now and then," he said.
    


In [15]:
import json
json_output = json.dumps(s_summary,indent=4)
print(json_output)

"\n    SOAP Notes:\n    Subjective: The patient is seen by a doctor in the back of a waiting room. The doctor says, \"Good morning, Ms.\n    Objective: Mr. Jones is the son of former U.S. President Harry S. Jones. Mr. Jones has been married to his wife, Barbara, for more than 40 years.\n    Assessment: Patient: Good morning, doctor. Doctor: How are you feeling today? Patient: I'm fine.\n    Plan: \"I\u2019m doing better, but I still have some discomfort now and then,\" he said.\n    "


In [16]:
json_output

'"\\n    SOAP Notes:\\n    Subjective: The patient is seen by a doctor in the back of a waiting room. The doctor says, \\"Good morning, Ms.\\n    Objective: Mr. Jones is the son of former U.S. President Harry S. Jones. Mr. Jones has been married to his wife, Barbara, for more than 40 years.\\n    Assessment: Patient: Good morning, doctor. Doctor: How are you feeling today? Patient: I\'m fine.\\n    Plan: \\"I\\u2019m doing better, but I still have some discomfort now and then,\\" he said.\\n    "'

**Structured Summary**

Subjective: The patient is seen by a doctor in the back of a waiting room. The doctor says, \"Good morning, Ms.\n

Objective: Mr. Jones is the son of former U.S. President Harry S. Jones. Mr. Jones has been married to his wife, Barbara, for more than 40 years.\n    

Assessment: Patient: Good morning, doctor. Doctor: How are you feeling today? Patient: I'm fine.\n    Plan: \"I\u2019m doing better, but I still have some discomfort now and then,\" he said.\n

Q1.How would you handle ambiguous or missing medical data in the transcript?

ANS - We use spaCy to extract key medical Entities from out Structured Summaery data.

q2.- What **pre-trained NLP models** would you use for medical summarization?

ANS -


In [17]:
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import Dataset, DataLoader
from transformers import AutoTokenizer, AutoModelForSequenceClassification, Trainer, TrainingArguments
from sklearn.metrics import accuracy_score, f1_score

Fine Tune ClinicalBert for Training Model

In [18]:
model_name = "emilyalsentzer/Bio_ClinicalBERT"

In [19]:
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForSequenceClassification.from_pretrained(model_name, num_labels=3)

config.json:   0%|          | 0.00/385 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/213k [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/436M [00:00<?, ?B/s]

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at emilyalsentzer/Bio_ClinicalBERT and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


In [20]:
new_text

' Physician: Good morning, Ms. Jones. How are you feeling today? Patient: Good morning, doctor. I’m doing better, but I still have some discomfort now and then. Physician: I understand you were in a car accident last September. Can you walk me through what happened? Patient: Yes, it was on September 1st, around 12:30 in the afternoon. I was driving from Cheadle Hulme to Manchester when I had to stop in traffic. Out of nowhere, another car hit me from behind, which pushed my car into the one in front.   Physician: That sounds like a strong impact. Were you wearing your seatbelt?   Patient: Yes, I always do.   Physician: What did you feel immediately after the accident?   Patient: At first, I was just shocked. But then I realized I had hit my head on the steering wheel, and I could feel pain in my neck and back almost right away.   Physician: Did you seek medical attention at that time?   Patient: Yes, I went to Moss Bank Accident and Emergency. They checked me over and said it was a whi

Here Fine tune of Bert


In [21]:
encoded_inputs = tokenizer(
    new_text,
    max_length=512,
    padding="max_length",
    truncation=True,
    return_tensors="pt"
)

In [22]:
encoded_inputs

{'input_ids': tensor([[  101,  7454,   131,  1363,  2106,   117,   182,  1116,   119,   179,
         18764,   119,  1293,  1132,  1128,  2296,  2052,   136,  5351,   131,
          1363,  2106,   117,  3995,   119,   178,   787,   182,  1833,  1618,
           117,  1133,   178,  1253,  1138,  1199, 19614,  1208,  1105,  1173,
           119,  7454,   131,   178,  2437,  1128,  1127,  1107,   170,  1610,
          4216,  1314, 14516,  6451,  5521,  3169,   119,  1169,  1128,  2647,
          1143,  1194,  1184,  2171,   136,  5351,   131,  4208,   117,  1122,
          1108,  1113, 14516,  6451,  5521,  3169,  2198,   117,  1213,  1367,
           131,  1476,  1107,  1103,  4427,   119,   178,  1108,  3759,  1121,
         22572, 12393,  1513,   177,  4654,  3263,  1106,  1299, 17824,  1165,
           178,  1125,  1106,  1831,  1107,  3404,   119,  1149,  1104,  8251,
           117,  1330,  1610,  1855,  1143,  1121,  1481,   117,  1134,  2873,
          1139,  1610,  1154,  1103,  

In [23]:
input_ids = encoded_inputs['input_ids']
attention_masks = encoded_inputs['attention_mask']

In [24]:
your_labels = [0]
labels = torch.tensor(your_labels)

In [25]:

from torch.utils.data import DataLoader, TensorDataset

train_data = TensorDataset(input_ids, attention_masks, labels)
train_dataloader = DataLoader(train_data, batch_size=8, shuffle=True)

for batch in train_dataloader:
    input_ids, attn_masks, labels = batch
    print(input_ids.shape)

torch.Size([1, 512])


In [26]:
from transformers import AdamW
from torch.optim.lr_scheduler import StepLR
from torch.nn import CrossEntropyLoss

optimizer = AdamW(model.parameters(), lr=1e-5)
scheduler = StepLR(optimizer, step_size=1, gamma=0.9)
loss_fn = CrossEntropyLoss()


In [27]:
from tqdm import tqdm
import torch

epochs = 3
for epoch in range(epochs):
    model.train()
    running_loss = 0.0
    correct_preds = 0
    total_preds = 0

    for batch in tqdm(train_dataloader):

        input_ids, attention_masks, labels = batch
        # Forward pass
        outputs = model(input_ids, attention_mask=attention_masks, labels=labels)
        loss = outputs.loss
        logits = outputs.logits

        # Backpropagation
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        # Calculate accuracy
        _, predicted = torch.max(logits, dim=1)
        correct_preds += (predicted == labels).sum().item()
        total_preds += labels.size(0)

        running_loss += loss.item()


    avg_loss = running_loss / len(train_dataloader)
    accuracy = correct_preds / total_preds
    print(f"Epoch {epoch+1}, Loss: {avg_loss:.4f}, Accuracy: {accuracy:.4f}")


    scheduler.step()

100%|██████████| 1/1 [00:05<00:00,  5.55s/it]


Epoch 1, Loss: 1.2922, Accuracy: 0.0000


100%|██████████| 1/1 [00:05<00:00,  5.31s/it]


Epoch 2, Loss: 1.1196, Accuracy: 0.0000


100%|██████████| 1/1 [00:04<00:00,  4.43s/it]

Epoch 3, Loss: 1.1560, Accuracy: 0.0000


In [29]:
def preprocess_text(text):
    inputs = tokenizer(text, padding='max_length', truncation=True, max_length=512, return_tensors="pt")
    return inputs


inputs = preprocess_text(new_text)


with torch.no_grad():
    outputs = model(**inputs)

logits = outputs.logits
predicted_class = torch.argmax(logits, dim=1).item()

label_map = {0: "Negative", 1: "Neutral", 2: "Positive"}
print("Sentiment:", label_map[predicted_class])


Sentiment: Neutral
